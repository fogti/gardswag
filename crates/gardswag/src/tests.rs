use super::*;

fn dflsubscr() -> impl tracing::subscriber::Subscriber {
    tracing_subscriber::fmt::Subscriber::builder()
        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
        .finish()
}

macro_rules! assert_interp {
    ($x:expr) => {{
        let result = crossbeam_utils::thread::scope(|s| main_interp(s, $x))
            .expect("unable to join threads");
        insta::assert_debug_snapshot!(result);
    }}
}

#[test]
fn chk_hello() {
    insta::assert_yaml_snapshot!(main_check(r#"std.stdio.write("Hello world!\n");"#).unwrap());
}

#[test]
fn chk_fibo() {
    tracing::subscriber::with_default(dflsubscr(), || {
        insta::assert_yaml_snapshot!(main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib
            "#
        )
        .unwrap());
    });
}

#[test]
fn run_fibo0() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 0
            "#,
        )
        .unwrap();
        assert_interp!(&x.0);
    });
}

#[test]
fn run_fibo1() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 1
            "#,
        )
        .unwrap();
        assert_interp!(&x.0);
    });
}

#[test]
fn run_fibo() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 5
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn chk_implicit_restr() {
    tracing::subscriber::with_default(dflsubscr(), || {
        insta::assert_yaml_snapshot!(main_check(
            r#"
                \x
                let id = \y y;
                .{
                  id;
                  x;
                  y = "{x}";
                }
            "#
        )
        .unwrap());
    });
}

#[test]
fn run_id() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let id = \x x;
                id 1
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn run_call_blti() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                std.plus 1 1
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn run_fix() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                let rec f = \a if (std.eq a 0) { 0 } { f 0 };
                f 1
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn run_update() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check(
            r#"
                .{
                  a = "what";
                  b = 1;
                  c = .{};
                } // .{
                  b = "no";
                  c = 50;
                }
            "#,
        )
        .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

#[test]
fn error_int_update() {
    tracing::subscriber::with_default(dflsubscr(), || {
        insta::assert_yaml_snapshot!(main_check("0//0").map_err(|e| e.to_string()));
    });
}

#[test]
fn run_ctrl_match() {
    tracing::subscriber::with_default(dflsubscr(), || {
        let x = main_check("match .this_is_a_variant 1 | .this_is_a_variant x => std.plus x 1")
            .unwrap();
        insta::assert_yaml_snapshot!(x);
        assert_interp!(&x.0);
    });
}

proptest::proptest! {
    #![proptest_config(proptest::test_runner::Config::with_cases(8192))]

    #[test]
    fn doesnt_crash(s in "[ -~]+") {
        if let Ok(x) = main_check(&s) {
            let _ = crossbeam_utils::thread::scope(|s| main_interp(s, &x.0))
                .expect("unable to join threads");
        }
    }
}
