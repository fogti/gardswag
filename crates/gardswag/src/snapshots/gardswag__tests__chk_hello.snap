---
source: crates/gardswag/src/tests.rs
expression: "main_check(r#\"std.stdio.write(\"Hello world!\\n\");\"#).unwrap()"
---
- - std
  - stdio
  - write
  - plus
  - minus
  - mult
  - eq
  - leq
  - not
  - spawn_thread
  - make_chan
  - send
  - recv
  - chan_send
  - chan_recv
  - None
  - Some
- stmts:
    - offset: 9
      inner:
        Call:
          prim:
            offset: 9
            inner:
              Dot:
                prim:
                  offset: 3
                  inner:
                    Dot:
                      prim:
                        offset: 0
                        inner:
                          Identifier: 1
                        extra:
                          ty:
                            Record:
                              2:
                                Record:
                                  3:
                                    Arrow:
                                      arg_multi: Linear
                                      arg:
                                        Literal: String
                                      ret:
                                        Literal: Unit
                              4:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    Literal: Int
                                  ret:
                                    Arrow:
                                      arg_multi: Linear
                                      arg:
                                        Literal: Int
                                      ret:
                                        Literal: Int
                              5:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    Literal: Int
                                  ret:
                                    Arrow:
                                      arg_multi: Linear
                                      arg:
                                        Literal: Int
                                      ret:
                                        Literal: Int
                              6:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    Literal: Int
                                  ret:
                                    Arrow:
                                      arg_multi: Linear
                                      arg:
                                        Literal: Int
                                      ret:
                                        Literal: Int
                              7:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    Var: 5
                                  ret:
                                    Arrow:
                                      arg_multi: Linear
                                      arg:
                                        Var: 6
                                      ret:
                                        Literal: Bool
                              8:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    Literal: Int
                                  ret:
                                    Arrow:
                                      arg_multi: Linear
                                      arg:
                                        Literal: Int
                                      ret:
                                        Literal: Bool
                              9:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    Literal: Bool
                                  ret:
                                    Literal: Bool
                              10:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    Arrow:
                                      arg_multi: Unrestricted
                                      arg:
                                        Literal: Unit
                                      ret:
                                        Literal: Unit
                                  ret:
                                    Literal: Unit
                              11:
                                Arrow:
                                  arg_multi: Erased
                                  arg:
                                    Literal: Unit
                                  ret:
                                    Record:
                                      12:
                                        ChanSend:
                                          Var: 7
                                      13:
                                        ChanRecv:
                                          Var: 7
                              14:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    Var: 8
                                  ret:
                                    Arrow:
                                      arg_multi: Linear
                                      arg:
                                        ChanSend:
                                          Var: 8
                                      ret:
                                        Literal: Bool
                              15:
                                Arrow:
                                  arg_multi: Linear
                                  arg:
                                    ChanRecv:
                                      Var: 9
                                  ret:
                                    TaggedUnion:
                                      16:
                                        Literal: Unit
                                      17:
                                        Var: 9
                          ident_multi: 0
                      key:
                        offset: 4
                        inner: 2
                        extra: ~
                  extra:
                    ty:
                      Record:
                        3:
                          Arrow:
                            arg_multi: Linear
                            arg:
                              Literal: String
                            ret:
                              Literal: Unit
                    ident_multi: 0
                key:
                  offset: 10
                  inner: 3
                  extra: ~
            extra:
              ty:
                Arrow:
                  arg_multi: Linear
                  arg:
                    Literal: String
                  ret:
                    Literal: Unit
              ident_multi: 0
          arg:
            offset: 16
            inner:
              FormatString:
                - offset: 17
                  inner:
                    PureString: "Hello world!\n"
                  extra:
                    ty:
                      Literal: String
                    ident_multi: 0
            extra:
              ty:
                Literal: String
              ident_multi: 0
      extra:
        ty:
          Literal: Unit
        ident_multi: 0
  term: ~
- forall: []
  forall_args: []
  ty:
    Literal: Unit

